File: apps/web/src/screens/UX-04-create-match/createMatchForm.test.ts
import { describe, expect, it } from "vitest";
import {
  canContinue,
  continueFromCreateMatch,
  initialFormState,
  resolveTemplateListState,
  templateOrganizationMismatch,
  type MatchTemplate,
} from "./createMatchForm";

describe("canContinue (UX-04 Validation)", () => {
  it("is disabled until a format is chosen", () => {
    const state = initialFormState("My Match", "GUEST");
    expect(canContinue(state)).toBe(false);
  });

  it("enables once a format is chosen", () => {
    const state = { ...initialFormState("My Match", "GUEST"), format: "T20" as const };
    expect(canContinue(state)).toBe(true);
  });
});

describe("templateOrganizationMismatch (UX-04 Validation)", () => {
  const templates: MatchTemplate[] = [{ id: "tpl-1", organizationId: "org-A", label: "Weekend T20" }];

  it("no mismatch when no template is selected", () => {
    const state = initialFormState("My Match", { organizationId: "org-A" });
    expect(templateOrganizationMismatch(state, templates)).toBe(false);
  });

  it("no mismatch for a guest match -- there is no organization context to mismatch against", () => {
    const state = { ...initialFormState("My Match", "GUEST"), selectedTemplateId: "tpl-1" };
    expect(templateOrganizationMismatch(state, templates)).toBe(false);
  });

  it("flags a template belonging to a different organization", () => {
    const state = { ...initialFormState("My Match", { organizationId: "org-B" }), selectedTemplateId: "tpl-1" };
    expect(templateOrganizationMismatch(state, templates)).toBe(true);
  });

  it("no mismatch when the template belongs to the current organization", () => {
    const state = { ...initialFormState("My Match", { organizationId: "org-A" }), selectedTemplateId: "tpl-1" };
    expect(templateOrganizationMismatch(state, templates)).toBe(false);
  });
});

describe("continueFromCreateMatch", () => {
  const fakeIdPort = { newId: () => "generated-id-1" };

  it("returns null when no format has been chosen -- Continue is disabled, so this must be unreachable via the UI, but the guard stands on its own", () => {
    const state = initialFormState("My Match", "GUEST");
    expect(continueFromCreateMatch(state, fakeIdPort)).toBeNull();
  });

  it("produces a DraftMatch with a client-generated id once a format is chosen", () => {
    const state = { ...initialFormState("My Match", "GUEST"), format: "T20" as const };
    const draft = continueFromCreateMatch(state, fakeIdPort);
    expect(draft).toEqual({
      id: "generated-id-1",
      matchLabel: "My Match",
      organizationId: null,
      format: "T20",
      templateId: null,
    });
  });

  it("resolves organizationId from ownership when not a guest match", () => {
    const state = { ...initialFormState("My Match", { organizationId: "org-A" }), format: "ODI_LIST_A" as const };
    const draft = continueFromCreateMatch(state, fakeIdPort);
    expect(draft?.organizationId).toBe("org-A");
  });
});

describe("resolveTemplateListState (UX-04 States/Error handling/Empty states)", () => {
  const templates: MatchTemplate[] = [{ id: "tpl-1", organizationId: "org-A", label: "Weekend T20" }];

  it("a successful fetch with results is populated, not from cache", () => {
    expect(resolveTemplateListState(true, templates, [])).toEqual({ status: "populated", templates, isFromCache: false });
  });

  it("a successful fetch with zero templates is the empty state", () => {
    expect(resolveTemplateListState(true, [], [])).toEqual({ status: "empty" });
  });

  // Error handling: "falls back to the last-synced local cache silently if any exist"
  it("a failed fetch falls back silently to the cache when one exists", () => {
    expect(resolveTemplateListState(false, null, templates)).toEqual({ status: "fetch-failed-fallback-to-cache", templates });
  });

  // Error handling: "otherwise offers 'Continue with custom setup' so the flow is never blocked"
  it("a failed fetch with no cache offers custom setup, never blocking the flow", () => {
    expect(resolveTemplateListState(false, null, [])).toEqual({ status: "fetch-failed-no-cache" });
  });
});

File: shared/src/commonTest/kotlin/com/kencric/scoring/ui/screens/ux04creatematch/CreateMatchFormTest.kt
package com.kencric.scoring.ui.screens.ux04creatematch

import com.kencric.scoring.core.ports.IdPort
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * `TASK-0041`, cross-platform parity (`C-7`): every case here mirrors
 * `apps/web/src/screens/UX-04-create-match/createMatchForm.test.ts`
 * (`TASK-0040`) input-for-input, expected-output-for-expected-output --
 * the same fixtures, not independently re-derived, so a genuine parity
 * gap between the two implementations would show up as one side's test
 * failing while the other's passes, not as two different tests neither
 * of which checks the same thing.
 */
class CreateMatchFormTest {

    private val fakeIdPort = object : IdPort {
        override fun newId(): String = "generated-id-1"
    }

    @Test fun canContinue_is_disabled_until_a_format_is_chosen() {
        val state = initialFormState("My Match", Ownership.Guest)
        assertFalse(canContinue(state))
    }

    @Test fun canContinue_enables_once_a_format_is_chosen() {
        val state = initialFormState("My Match", Ownership.Guest).copy(format = MatchFormat.T20)
        assertTrue(canContinue(state))
    }

    private val templates = listOf(MatchTemplate(id = "tpl-1", organizationId = "org-A", label = "Weekend T20"))

    @Test fun templateOrganizationMismatch_none_when_no_template_selected() {
        val state = initialFormState("My Match", Ownership.Organization("org-A"))
        assertFalse(templateOrganizationMismatch(state, templates))
    }

    @Test fun templateOrganizationMismatch_none_for_a_guest_match() {
        val state = initialFormState("My Match", Ownership.Guest).copy(selectedTemplateId = "tpl-1")
        assertFalse(templateOrganizationMismatch(state, templates))
    }

    @Test fun templateOrganizationMismatch_flags_a_different_organization() {
        val state = initialFormState("My Match", Ownership.Organization("org-B")).copy(selectedTemplateId = "tpl-1")
        assertTrue(templateOrganizationMismatch(state, templates))
    }

    @Test fun templateOrganizationMismatch_none_when_organizations_match() {
        val state = initialFormState("My Match", Ownership.Organization("org-A")).copy(selectedTemplateId = "tpl-1")
        assertFalse(templateOrganizationMismatch(state, templates))
    }

    @Test fun continueFromCreateMatch_returns_null_when_no_format_chosen() {
        val state = initialFormState("My Match", Ownership.Guest)
        assertNull(continueFromCreateMatch(state, fakeIdPort))
    }

    @Test fun continueFromCreateMatch_produces_a_draft_with_a_client_generated_id() {
        val state = initialFormState("My Match", Ownership.Guest).copy(format = MatchFormat.T20)
        val draft = continueFromCreateMatch(state, fakeIdPort)
        assertEquals(
            DraftMatch(id = "generated-id-1", matchLabel = "My Match", organizationId = null, format = MatchFormat.T20, templateId = null),
            draft,
        )
    }

    @Test fun continueFromCreateMatch_resolves_organizationId_when_not_a_guest_match() {
        val state = initialFormState("My Match", Ownership.Organization("org-A")).copy(format = MatchFormat.ODI_LIST_A)
        val draft = continueFromCreateMatch(state, fakeIdPort)
        assertEquals("org-A", draft?.organizationId)
    }

    @Test fun resolveTemplateListState_successful_fetch_with_results_is_populated_not_from_cache() {
        assertEquals(
            TemplateListState.Populated(templates, isFromCache = false),
            resolveTemplateListState(fetchSucceeded = true, fetchedTemplates = templates, cachedTemplates = emptyList()),
        )
    }

    @Test fun resolveTemplateListState_successful_fetch_with_zero_templates_is_empty() {
        assertEquals(
            TemplateListState.Empty,
            resolveTemplateListState(fetchSucceeded = true, fetchedTemplates = emptyList(), cachedTemplates = emptyList()),
        )
    }

    // Error handling: "falls back to the last-synced local cache silently if any exist."
    @Test fun resolveTemplateListState_failed_fetch_falls_back_silently_to_cache() {
        assertEquals(
            TemplateListState.FetchFailedFallbackToCache(templates),
            resolveTemplateListState(fetchSucceeded = false, fetchedTemplates = null, cachedTemplates = templates),
        )
    }

    // Error handling: "otherwise offers 'Continue with custom setup' so the flow is never blocked."
    @Test fun resolveTemplateListState_failed_fetch_with_no_cache_never_blocks_the_flow() {
        assertEquals(
            TemplateListState.FetchFailedNoCache,
            resolveTemplateListState(fetchSucceeded = false, fetchedTemplates = null, cachedTemplates = emptyList()),
        )
    }
}
